psycopg/connection_type.c

```diff
@@ -110,27 +110,44 @@ psyco_conn_cursor(connectionObject *self, PyObject *args, PyObject *kwargs)
     if (PyObject_IsInstance(obj, (PyObject *)&cursorType) == 0) {
         PyErr_SetString(PyExc_TypeError,
             "cursor factory must be subclass of psycopg2.extensions.cursor");
+        Py_DECREF(obj);
+        obj = NULL;
         goto exit;
     }
 
-    if (0 > curs_withhold_set((cursorObject *)obj, withhold)) {
-        goto exit;
+    curs = (cursorObject *)obj;
```

Comment:
Here we have the same object assigned to both `curs` and `obj` so making sure we decref everything alright is tricky. I'm writing a patch for it.